File: src/main/kotlin/magicbook/gtlitecore/api/recipe/builder/MinimumHeightRecipeBuilder.java
package magicbook.gtlitecore.api.recipe.builder;

import gregtech.api.recipes.Recipe;
import gregtech.api.recipes.RecipeBuilder;
import gregtech.api.recipes.RecipeMap;
import gregtech.api.util.EnumValidationResult;
import magicbook.gtlitecore.api.recipe.property.MinimumHeightProperty;
import magicbook.gtlitecore.api.utils.GTLiteLog;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.jetbrains.annotations.NotNull;

public class MinimumHeightRecipeBuilder extends RecipeBuilder<MinimumHeightRecipeBuilder>
{

    public MinimumHeightRecipeBuilder() {}

    public MinimumHeightRecipeBuilder(Recipe recipe,
                                      RecipeMap<MinimumHeightRecipeBuilder> recipeMap)
    {
        super(recipe, recipeMap);
    }

    public MinimumHeightRecipeBuilder(MinimumHeightRecipeBuilder recipeBuilder)
    {
        super(recipeBuilder);
    }

    @Override
    public MinimumHeightRecipeBuilder copy()
    {
        return new MinimumHeightRecipeBuilder(this);
    }

    @Override
    public boolean applyPropertyCT(@NotNull String key, @NotNull Object value)
    {
        if (key.equals(MinimumHeightProperty.KEY))
        {
            this.minHeight(((Number) value).intValue());
            return true;
        }
        return super.applyPropertyCT(key, value);
    }

    public MinimumHeightRecipeBuilder minHeight(int height)
    {
        if (height <= -64 || height >= 256) // TODO Should we consider the special situation when player install a height-extended mod?
        {
            GTLiteLog.logger.error("Minimum Height cannot be less than -64 or greater than 256", new IllegalArgumentException());
            this.recipeStatus = EnumValidationResult.INVALID;
        }
        this.applyProperty(MinimumHeightProperty.getInstance(), height);
        return this;
    }

    public int getMinHeight()
    {
        return this.recipePropertyStorage == null ? 0 : this.recipePropertyStorage.get(MinimumHeightProperty.getInstance(), 0);
    }

    public String toString()
    {
        return new ToStringBuilder(this)
                .appendSuper(super.toString())
                .append(MinimumHeightProperty.getInstance().getKey(), getMinHeight())
                .toString();
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/recipe/builder/MinimumHeightRecipeBuilder.kt
package gregtechlite.gtlitecore.api.recipe.builder

import gregtech.api.recipes.Recipe
import gregtech.api.recipes.RecipeBuilder
import gregtech.api.recipes.RecipeMap
import gregtech.api.util.EnumValidationResult
import gregtechlite.gtlitecore.api.LOGGER
import gregtechlite.gtlitecore.api.extension.buildToString
import gregtechlite.gtlitecore.api.recipe.GTLiteRecipeProperties

class MinimumHeightRecipeBuilder : RecipeBuilder<MinimumHeightRecipeBuilder>
{
    val minHeight
        get() = recipePropertyStorage?.let { recipePropertyStorage.get(GTLiteRecipeProperties.MINIMUM_HEIGHT, 0) } ?: 0

    constructor()

    @Suppress("unused")
    constructor(recipe: Recipe, recipeMap: RecipeMap<MinimumHeightRecipeBuilder>) : super(recipe, recipeMap)

    constructor(recipeBuilder: MinimumHeightRecipeBuilder) : super(recipeBuilder)

    override fun copy(): MinimumHeightRecipeBuilder = MinimumHeightRecipeBuilder(this)

    override fun applyPropertyCT(key: String, value: Any): Boolean
    {
        if (key == GTLiteRecipeProperties.MINIMUM_HEIGHT.key)
        {
            minHeight((value as Number).toInt())
            return true
        }
        return super.applyPropertyCT(key, value)
    }

    fun minHeight(height: Int): MinimumHeightRecipeBuilder = apply {
        if (height <= -64 || height >= 256)
        {
            LOGGER.error("Minimum Height cannot be less than -64 or greater than 256", IllegalArgumentException())
            recipeStatus = EnumValidationResult.INVALID
        }
        applyProperty(GTLiteRecipeProperties.MINIMUM_HEIGHT, height)
    }

    override fun toString(): String = buildToString {
        appendSuper(super.toString())
        append(GTLiteRecipeProperties.MINIMUM_HEIGHT.key, minHeight)
    }
}
